ui/org.eclipse.pde.core/src/org/eclipse/pde/internal/core/target/VirtualArtifactRepository.java

```diff
@@ -0,0 +1,199 @@
+	// BundleInfo or IFeatureModel
+	private final Map<IArtifactDescriptor, Object> artifacts = new HashMap<>();
+
+	public VirtualArtifactRepository(IProvisioningAgent agent, ITargetLocation targetLocation) {
+		super(agent, NAME + getLocationSafe(targetLocation), targetLocation.getType(), null,
+				URI.create(MEMORY_PREFIX + UUID.randomUUID()), DESCRIPTION + '\n' + targetLocation.serialize(), null,
+				null);
+		Assert.isTrue(targetLocation.isResolved());
+		for (TargetBundle targetBundle : targetLocation.getBundles()) {
+			if (!targetBundle.getStatus().isOK()) {
+				PDECore.log(Status.warning(NLS.bind(Messages.VirtualArtifactRepository_0, targetBundle)));
+				continue;
+			}
+			BundleInfo bundleInfo = targetBundle.getBundleInfo();
+			IArtifactKey artifactKey = BundlesAction.createBundleArtifactKey(bundleInfo.getSymbolicName(),
+					bundleInfo.getVersion());
+			IArtifactDescriptor artifactDesriptor = new ArtifactDescriptor(artifactKey);
+			artifacts.put(artifactDesriptor, bundleInfo);
+		}
+		for (TargetFeature targetFeature : targetLocation.getFeatures()) {
+			IArtifactKey artifactKey = FeaturesAction.createFeatureArtifactKey(targetFeature.getId(),
+					targetFeature.getVersion());
+			IArtifactDescriptor artifactDesriptor = new ArtifactDescriptor(artifactKey);
+			artifacts.put(artifactDesriptor, targetFeature.getFeatureModel());
+		}
+	}
+
+	private static String getLocationSafe(ITargetLocation targetLocation) {
+		try {
+			return targetLocation.getLocation(false);
+		} catch (CoreException e) {
+			return "<unknown>"; //$NON-NLS-1$
+		}
+	}
+
+	@Override
+	public IStatus getRawArtifact(IArtifactDescriptor descriptor, OutputStream destination, IProgressMonitor monitor) {
+		Object artifactModel = artifacts.get(descriptor);
+		if (artifactModel == null) {
+			return Status.error(NLS.bind(Messages.VirtualArtifactRepository_1, descriptor));
+		}
+		try {
+			transferArtifact(artifactModel, destination);
+			return Status.OK_STATUS;
+		} catch (Exception e) {
+			return Status.error(e.getLocalizedMessage(), e);
+		}
+	}
+
+	private void transferArtifact(Object artifactModel, OutputStream destination) throws Exception {
+		if (artifactModel instanceof BundleInfo bundleInfo) {
+			URI location = bundleInfo.getLocation();
+			if (location == null) {
+				throw new FileNotFoundException(bundleInfo.getSymbolicName());
+			}
+			try (InputStream is = location.toURL().openStream()) {
+				is.transferTo(destination);
+			}
+		} else if (artifactModel instanceof IFeatureModel featureModel) {
+			String installLocation = featureModel.getInstallLocation();
+			if (installLocation != null) {
+				File featureJar = new File(installLocation);
+				if (featureJar.isFile()) {
+					Files.copy(featureJar.toPath(), destination);
+					return;
+				}
+			}
+			IFeature feature = featureModel.getFeature();
+			// Generate in-memory feature jar (with only the feature.xml)
+			JarOutputStream jos = new JarOutputStream(destination);
```

Comment:
What happend to the try/with resources here?

Reply:
Closing the JarOutputStream also closes the underlying `destination` stream. Given that this stream is managed by the caller, it mustn't be closed by this method.

Reply:
you are right, what a mess :-D